fuzz/provider.c

```diff
@@ -0,0 +1,655 @@
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_MAC_update(ctx, (unsigned char *) text, sizeof(text))) {
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_MAC_final(ctx, buf, &final_l, sizeof(buf))) {
+        r = 0;
+        goto end;
+    }
+
+end:
+    EVP_MAC_CTX_free(ctx);
+    return r;
+}
+
+static int do_evp_rand(EVP_RAND *evp_rand, const OSSL_PARAM params[])
+{
+    int r = 1;
+    EVP_RAND_CTX *ctx = NULL;
+    unsigned char buf[4096];
+
+    if (!(ctx = EVP_RAND_CTX_new(evp_rand, NULL))) {
+        r = 0;
+        goto end;
+    }
+
+    if (EVP_RAND_CTX_set_params(ctx, params) <= 0) {
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_RAND_generate(ctx, buf, sizeof(buf), 0, 0, NULL, 0)) {
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_RAND_reseed(ctx, 0, 0, 0, NULL, 0)) {
+        r = 0;
+        goto end;
+    }
+
+end:
+    EVP_RAND_CTX_free(ctx);
+    return r;
+}
+
+static int do_evp_sig(EVP_SIGNATURE *evp_sig, const OSSL_PARAM params[])
+{
+    return 0;
+}
+
+static int do_evp_asym_cipher(EVP_ASYM_CIPHER *evp_asym_cipher, const OSSL_PARAM params[])
+{
+    return 0;
+}
+
+static int do_evp_kem(EVP_KEM *evp_kem, const OSSL_PARAM params[])
+{
+    return 0;
+}
+
+static int do_evp_key_exch(EVP_KEYEXCH *evp_kdf, const OSSL_PARAM params[])
+{
+    return 0;
+}
+
+static int do_evp_md(EVP_MD *evp_md, const OSSL_PARAM params[])
+{
+    int r = 1;
+    unsigned char md_value[EVP_MAX_MD_SIZE];
+    unsigned int md_len;
+    EVP_MD_CTX *mdctx = NULL;
+
+    if (!(mdctx = EVP_MD_CTX_new())) {
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_MD_CTX_set_params(mdctx, params)) {
+        r = 0;
+        goto end;
+    }
+
+    if (!EVP_DigestInit_ex2(mdctx, evp_md, NULL)) {
+        r = 0;
+        goto end;
+    }
+    if (!EVP_DigestUpdate(mdctx, "Test", strlen("Test"))) {
+        r = 0;
+        goto end;
+    }
+    if (!EVP_DigestFinal_ex(mdctx, md_value, &md_len)) {
+        r = 0;
+        goto end;
+    }
+
+end:
+    EVP_MD_CTX_free(mdctx);
+    return r;
+}
+
+#define EVP_FUZZ(source, evp, f) \
+    do { \
+        evp * alg = sk_##evp##_value(source, *algorithm % sk_##evp##_num(source)); \
+        OSSL_PARAM *fuzzed_params; \
+        \
+        if (!alg) { \
+            break; \
+        } \
+        fuzzed_params = fuzz_params((OSSL_PARAM*) evp##_settable_ctx_params(alg), &buf, &len); \
+        if (fuzzed_params) { \
+            f(alg, fuzzed_params); \
+        } \
+        OSSL_PARAM_free(fuzzed_params); \
+    } while (0);
+
+static uint64_t DFLT_OP = 0;
+static int64_t DFLT_ALG = 0;
+
+int FuzzerTestOneInput(const uint8_t *buf, size_t len)
+{
+    int r = 1;
+    uint64_t *operation = &DFLT_OP;
+    int64_t *algorithm = &DFLT_ALG;
+
+    if (!read_uint(&buf, &len, &operation)) {
+        r = 0;
+        goto end;
+    }
+
+    if (!read_int(&buf, &len, &algorithm)) {
+        r = 0;
+        goto end;
+    }
+
+    switch (*operation % 10) {
```

Comment:
I suggest you just make that *operation. Don't see the point of %10, this will just confuse the fuzzers.

Reply:
I can certainly change that to `*operation`, but I'm not certain which one would perform better. Removing the module might cause more test cases to fail.

My gut feeling is that it would probably not matter in the long run.